src/hashmap/concurrent_unordered_map.cuh

```diff
@@ -0,0 +1,691 @@
+    {
+        return m_current == other.m_current && m_begin == other.m_begin && m_end == other.m_end;
+    }
+    
+    __host__ __device__ reference& operator*()
+    {
+        return *m_current;
+    }
+    
+    __host__ __device__ const reference& operator*() const
+    {
+        return *m_current;
+    }
+
+    __host__ __device__ const pointer operator->() const
+    {
+        return m_current.operator->();
+    }
+    
+    __host__ __device__ pointer operator->()
+    {
+        return m_current;
+    }
+    
+private:
+    iterator_type m_current;
+    iterator_type m_begin;
+    iterator_type m_end;
+};
+
+template <class T>
+__host__ __device__ bool operator==(const cycle_iterator_adapter<T>& lhs, const cycle_iterator_adapter<T>& rhs)
+{
+    return lhs.equal(rhs);
+}
+
+template <class T>
+__host__ __device__ bool operator!=(const cycle_iterator_adapter<T>& lhs, const cycle_iterator_adapter<T>& rhs)
+{
+    return !lhs.equal(rhs);
+}
+
+/**
+ * Does support concurrent insert, but not concurrent insert and probping.
+ *
+ * TODO:
+ *  - add constructor that takes pointer to hash_table to avoid allocations
+ *  - extend interface to accept streams
+ */
+template <typename Key,
+          typename Element,
+          Key unused_key,
+          typename Hasher = default_hash<Key>,
+          typename Equality = equal_to<Key>,
+          typename Allocator = managed_allocator<thrust::pair<Key, Element> >,
+          bool count_collisions = false>
+class concurrent_unordered_map : public managed
+{
+
+public:
+    typedef size_t                                          size_type;
+    typedef Hasher                                          hasher;
+    typedef Equality                                        key_equal;
+    typedef Allocator                                       allocator_type;
+    typedef Key                                             key_type;
+    typedef thrust::pair<Key, Element>                      value_type;
+    typedef Element                                         mapped_type;
+    typedef cycle_iterator_adapter<value_type*>             iterator;
+    typedef const cycle_iterator_adapter<value_type*>       const_iterator;
+
+private:
+    union pair2longlong
+    {
+        unsigned long long int  longlong;
+        value_type              pair;
+    };
+    
+public:
+
+    explicit concurrent_unordered_map(size_type n,
+                                      const mapped_type unused_element,
+                                      const Hasher& hf = hasher(),
+                                      const Equality& eql = key_equal(),
+                                      const allocator_type& a = allocator_type())
+        : m_hf(hf), m_equal(eql), m_allocator(a), m_hashtbl_size(n), m_hashtbl_capacity(n), m_collisions(0), m_unused_element(unused_element)
+    {
+        m_hashtbl_values = m_allocator.allocate( m_hashtbl_capacity );
```

Comment:
Is this information that we can precompute with a "dry" non process run so we can tell our code on the outside how to handle these allocations as opposed to doing them here. The issue I am finding is that right now if I want to start calculating worse possible scenarios for allocation. e.g. if count where = 1 for every group, I have to go to several places to find that out. And that information is important if we are trying to build systems that will approach full utilization of resources. Being able to ask an api, how much memory do you need, ok let me allocate that for you. Is much safer and more maintainable in the long run.